### PR TITLE
Maintain column indentation for list of functions

### LINKED ISSUE #2158: List of functions misaligned and breaks (Elmish)
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKBbBBXANosAwrOgA4DOAOgE4UB21dsJitYAYgIYBGeAdAEqIO0AC68AClSaUatJi3a50HBvVl4RYMohEiAlrQDmZALIssWnfqNkwWEvA4jEAZSsHjYALyrVECAD0AWAuIgCeeGAAjH7%2BGmAkUuTesf7%2B-BzwerC8AJK0JFhiktIA2gQAFojQANaI8GAAZhw42qlpHZ1dYADytJUcRogAFI1YrAD6YAC0AHyWuh62AD7z9o7Obos2AJQAur6yaUEh4ZEATO3xidIpR10ZWTn5hcVJZKXt3ZXVdQ3NrUQXy6fQGQ1G4zAUzmC2snlWdgcTlc7l2B1k7Ru5EgABpwFBtNtjKgwKV8RBQJ1IAAScLMEmQAwiXHk-w0gBuLSwyDQZO6AAY8d1IPkkLQRC49AAvZBCroAFlZED2rIAvnKKUqoLSwvS0IzxSyqdTOThuSS%2BV0ohqOpATBwAB4AGQMiCdLEMIgqRqt535WpVaXVrMptp1eqgLHgPUaLtorjOsq1HK5PNJWog5xtaUgAFFaNHY66fVScI1IAG1TbQznw2nIFxYLAcCW2SbUxaM2AAMzZtn5bRUERsAwtAByiAA7jhi32ICIqNzK0Hq8m6wzG83WxAU2a05bOvK51A3Og9HAcLBaABBET5wtxpPdAHaZf%2BYNpGttun1zctvu7uavJdgArMekAuCQwiIAAQogjSwFQiDiBwVAcOgOiIFQ27%2BAuS6dIG76rsaP4bk2-5rqaQHpt0ABs4GQdBcEIUhTqwJOWHQBw2j5OysBcfoV44fOi6IG%2BEAfv4X47uu%2Bp-jhgH7l2ADsDFQdAsHwYhiAAKokMwVBcTxtB8QJ2S0MJTQtK%2BBFViGa6kXJ5EKe2e6dt0AAcalMVpSEEDg3FkEQtBkHhoiIZZL5ibZK72SRuq-s5AGudRB4dAAnN5GnMdpZjoFwWGRdZ0UdIREnEWGjlQPJyVUUp3RRIKXYQepmksYgRCXhZx5ReJYCSZq8URg2SWUR2wENdazWMRp16NM4VBEOgyiWXhJVpGV-UVbWVUjVutXjTRvpZW12mnuezZCT1xV9QNYDSdqu01WNbkTVavbTa1c0LedF5XV2a23dt34JWR%2B0valXZREezWiiwIh9AAKlQYQAOp6F6RWAkDcWVaDTng8adXuVaYGfdB15SOM8AACKIDOZ4Latok45%2BDn41ATIucTb2dFE9HNfaDq5I0iNVLQuaAi4FSISI6PwJjx7yv6MVEbjO0cwazIHa9R186pguOvkjR6A6PQGU4iG5g6iSIGQZDmZZIEq6Vdls0N9ZczrkMNV5hsOoIcBUPA8uK12yus1J7PDV7EP1VamX%2B4HiHwGOWD5VhMa5M46BkJZUSR4NePDeguD6DO8ZsIhyi6FhiOg978edH64HJ8HJhl3oFeIFXVA10z4HQBUqHCAtEyTnoCveoX93R57hqNyTzdTcKQuU2hYQ9FQLqhaH3pKy7G1u1HHsMrHROHWlaTnFm-vrxwm-b3ooVpxnVBZznefHgXqvlerIMl07t3Xu-c64NzjkvDo5wPrCnvo-HeIgO44HLq6EBTgB7NSHiPUQWFx6T0xjPB6NJdrnzDDzPWUCYar0dAANVTDBAwWQjB70sh5Q%2B-hNp3SIbJTmC8IG8ygWTahDo2DjFEOZBhBYPAsIPoQueZ8%2BEX11lffw5wBbCJprAEQABxHQ1tbb23MjI0C7DlTHyLhrYaz0lE%2By6OcA2q8gGuhgpeWoME0K1B0GQPoLh0IdWbOnbqXZeq-y2v-GST1Ro2KblAv2woJzTmcToDiLB67MDpibWgGNzLXgLHlAqVAv7BJuqErh8iCYUWiZA6%2BidhQAGlECIBICLMWLB8i%2BIwo%2BLGNlXaxXdsXee2t%2BEULSN2JqwjJZnjIBUYx3QI6lOBhEzW1iyGXy7N2FeXQIIeDwJTQw6d4ao0QFwEwCAnxdBCb0tW-TLGJUJqs5R6zb6wJnIYWgojaDiKvJKN5TgsBIURrAOG4onCO2utjBZ4THrLKiQ82xnRuwwK2deZBWFaCgvZG6K8hh8lYQyQYbJV4inPhKVcv%2BNyAF3MqXCmJoyqFbKQfoR8Tp0JcEcP5WADsjAJO7t09aHDzGz1PhU7mazujdiEVsmmz9uB4Emc-CoLgwggodHyuRwrqqwtrOQlREBuzqK2Q0ppwKRD5HcYMIearIUUqWVYrVbYdXrIcVslxgwaiPjIJTWA1MJyhXqIyrurp9FIUMUS5m%2BEyVhJtdCu19ztViq6N2OJLrUI5SQtKsg0AqB6DPOi5w8AdJZKvHTaAAU0KCSCSSiFkaykar2tS%2BNjzxW1K2aEbNohTlICtTW-Eew5SMjIGwB2JJeqqiAA)

#### Code

```fsharp
module Comps

open Fable.React.Props
open Fulma

let settingsMenu settings updateSettings =

    // Style 1
    let props =
        Radio.Input.Props[Checked false
                          OnChange(fun _ -> settings |> updateSettings)]

    // Style 2
    let props =
        Radio.Input.Props[
            Checked false
            OnChange(fun _ -> settings |> updateSettings)]

    props
```
            

#### Result

```fsharp
module Comps

open Fable.React.Props
open Fulma

let settingsMenu settings updateSettings =

    // Style 1
    let props =
        Radio.Input.Props[Checked false
        OnChange(fun _ -> settings |> updateSettings)]

    // Style 2
    let props =
        Radio.Input.Props[Checked false
        OnChange(fun _ -> settings |> updateSettings)]

    props

```
            
#### Problem description

This is an extraction of some involved Elmish code. 

`Fantomas.FormatConfig+FormatException: Formatted content is not valid F# code`

The online tool clarifies the issue:

```
(11,8) (11, 53)Error597parse
Successive arguments should be separated by spaces or tupled, and arguments involving function or method applications should be parenthesized

(16,8) (16, 53)Error597parse
Successive arguments should be separated by spaces or tupled, and arguments involving function or method applications should be parenthesized
```

It seems as if the top level bind is being prioritised over the list constructions.

The issue seems to have started somewhere between 4.6.0 and 4.7.x

#### Extra information

- [x] The formatted result breaks my code.
- [x] The formatted result gives compiler warnings.
- [x] I or my company would be willing to help fix this.

#### Options

Fantomas master branch at 2022-03-15T07:45:16Z - f10e2392ff7207059345cf75223ef9431c36821d

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#ignore-files-fantomasignore)?</sub>

### PR DESCRIPTION
**Pull request by cgravill**
Fixes #2158 

I wasn't sure about the added duplication in `IndexWithoutDotExpr` but factoring the change seemed to add as much complexity. Elsewhere seems to favour mild duplication. Happy to adjust as preferred.

Noticed in the ListTests.fs file the last function isn't marked as a test. I've checked and it runs fine. I have NOT added it to this PR but happy to or add a separate PR to get properly added and running:
 https://github.com/fsprojects/fantomas/blob/2d5db3e6c7b7fb47e374505765ed530dfd934aa2/src/Fantomas.Core.Tests/ListTests.fs#L2302-L2321

### CODE REVIEW

File: src/Fantomas.Core/CodePrinter.fs [hunk 1]
```diff
@@ -1401,10 +1401,17 @@ and genExpr astContext synExpr ctx =
             )
 
         | IndexWithoutDotExpr (identifierExpr, indexExpr) ->
-            genExpr astContext identifierExpr
-            +> sepOpenLFixed
-            +> genExpr astContext indexExpr
-            +> sepCloseLFixed
+            match indexExpr with
+            | Sequential _ ->
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1405]: I think it is *ok* to always have the sequential case when the expression is multiline.
This problem isn't necessarily related to `SynExpr.Sequential`, more general to multiline expressions. [Example](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMBDAMrEcSQHYDm%2BAlgM44C8YA2mAAxgDcYAjGALoAdAE4DSw0ZhwAzKkLph640hAhSArsoAeYALQA%2BJSqN5CJMxWp0Wh47YDKsALaIAwulw0aAOjtZcWKmgAWURHACNEIQAKG1s4sAB6BLByIngwOEdnUixY%2BONcIXJuABo8-JVC8gAmMpFlCuN1ZQBPXQN6xvyksHKulTaAah4%2B2wBKLwAxIlH%2BSBLwKBpsAIoaVGZFiFBbSAASLBaAB2Q0SCoc%2Ba2VfYA3XHQ1U8341gX4yABJUiQcuyoAF7Id5xAAs1wgfGuAF8QdsIVADsdnudLnCbnt7o9nkwERBSnioEFcJpqKREPhEBQsAALK7xbjVVgIqFGWHXHbGfaHE4bSBU%2BAAeSkZMQvhamHpXMxDyeG1x8VqhMgAFEfsLRVLdugpJAWTD0ZyjNzkXywkR0FqMVi5WgFXEAMzo43fZZCLCTC4PAByiAA7ugLsDCVghE99WzDQiTbyzubYJbnYibTjCaCkxBIHYjrhoIgAEKIKSwISIAAKhVwziwkStRlD4dsrJU7KMRoxPJR8cT0Zl2PlhIArBmoNncwWiyWKbA-ZFoB5EN9brB5wFYKQ6yoG4gIy2o7skbGoN3N3dZan4gA2EdZnN5wvF0sAVSOJyE8%2BWS5X-io683qgeZZdwgVsVHbTNDy7C1Tz7W0XjiAB2G8x3vSdSzcDwaBcdc6DDaAsBLf8wCkQCdybA0OV7TszWgpMz37O1CQADmQu8J0fRAQnCWsRxI9AgPIyNKIPai41o3sUwHeIAE5WPHB8p2w9A-140jgLAUD4RE00xITGDJMYhk3mVFDEAAQSkGshGwrJcCI7d1M0sBwMRUTj3Eg8DPg2xuAJD5TIU0s7FCQIExUwk%2BIE4xmxA-dpTcyATzo2CLziRk5LzCyrOCxxQuUjcRwcwS92E%2BKdPcvTkq8%2B0fKdEy2LMoRYA0eAABFEEDXKrPssMyOiii2yo8rUSwfTzyktL02VYlNE%2BKQABUaSpOwaRLLAAHUqHgWkiOZYrYtK41IL5C5Rqq8bDLS4dppJObFqpFV%2BLFVb3U27a6RHS89v6oTBu0o8RrGhjvOMbhrxu2bSBkTRBTffwSxVTQjlLTxfwK5jvqMGKNLio6EtOoG4Jq0GkIhgAlRA4CEeA3p2kdQUxlRsaclyYxRAnzuB4mjG4Fjycpkt4G9NRuKEYVPhrRx1hHbhHNxjthscNR0ACQNyWmIRHH8Kz5tNTmicJbhZOVCmqfgIJldVoMNa1rBupvaAaUrfDIgAfT9Ladrlw6FYBjmJIukGjCZG8ZsaoRcBaQUhEsDbPY%2BjHvb%2Bsq-bRAOucJao-LiSAw6ECOo5j2gsGF0Xxcl6XDaTsChoBpWVaoNXEBt7XIl13l9dS2xqiVD5w8j6PY4thum5bu2eOVR3nas93471facZ9iD8bTzzA%2B5lRqjqj4ZoANXPfMLngC5yFphP4iYxnIQGmv-vZ1fpWqzOpp3klJg0fC0cPn4T7PoiGerlpFO98zrpwNoqa6r9NCtVgFgAA4tgRGyNECo3XH-Ecl9AHOVrlBSqYCu7GGqODHeltG5BnzMpaAABrfMEdqHYBoIKUgdgqyuATCLdG8RIp9SxjfIBeNhpJXwRNbupMPi%2BgDOQ7As4qTt0QO1GQpAqBrlIGZH4XEIhyB6o2H6JVk4CIBkIteGdFR8z7oGcgpB36kE-uuf4lj-BqFLPNWA3xfh%2BBUURbhWDWbHV0j2Yx4C4jVGNuYqypAfy3GnBQDRkQFEXGUWjSuXC1ILxZjgmieDAkEKMA6YyJCG6inwFWMI8BcBuFgDQE%2BEim5eNSbog6%2Bjfa4ICY-dehIHTZ12EgyIVBsh%2BHQAAaUQIgI4biqRYG%2BLQ3ANjz5xG8Wk%2BWy9BEeTaSYx0vcc4UJmVQ0UNBGrNR%2BL6OgiBzakKbkglGVScLaJ4UzPh2C76ZNaUdJ%2B8QHTby2YUQK8jaDQCEH0r0NZ4BPiUeudq0B3AR08apJ6PiMn%2BMJjklQDoX45x6QC-pDxfBNTULhNQRxxSSlhVFXhv1b7AJOg-V57T3mQJzsM0ZM1S6aOFNs0guygzJLSqwK%2B-AHm%2BISkYtZQTbAOmITnHFgQsBBAQMGFJcK0mLD4CCc4NBJhVI2N46EQA)

You could do something like:
```fsharp
let genIndexExpr = genExpr astContext indexExpr

 genExpr astContext identifierExpr
  +> sepOpenLFixed
  +> expressionFitsOnRestOfLine genIndexExpr  (atCurrentColumnIndent genIndexExpr)
  +> sepCloseLFixed
```
 
People shouldn't write multiline indexer expressions in the first place, so I'm not overly bothered by the result of these.

File: src/Fantomas.Core.Tests/ListTests.fs [hunk 2]
```diff
@@ -2319,3 +2319,22 @@ let x: double[][] = [ [ 1.0, 2.0, 3.0 ] ]
 let foo (x: double[]) (y: object[][]) : string[,] = x :> int[]
 let foo<'T> (x: 'T[]) = x
 """
+
+[<Test>]
+let ``Maintain indentation on function list, 2158`` () =
```
Comment (nojaf) [inline src/Fantomas.Core.Tests/ListTests.fs:2324]: Could you lower-case `Maintain` please?
https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#unit-test

**Review by nojaf — CHANGES_REQUESTED** [review]
Many thanks, @cgravill for picking this up!

Would you mind adding in a Changelog entry?
https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#changelog
You will probably need to add the `Unreleased` header section.
And please include all duplicate issues in the changelog:
- https://github.com/fsprojects/fantomas/issues/2195
- https://github.com/fsprojects/fantomas/issues/2201
- https://github.com/fsprojects/fantomas/issues/2242
- https://github.com/fsprojects/fantomas/issues/2392

There is no need to add tests for these duplicates as people should have added a space in the first place. But I'm considering all these closed and it should be recorded in the change log.

File: src/Fantomas.Core/CodePrinter.fs (hunk 1 above)
Comment (cgravill) [reply src/Fantomas.Core/CodePrinter.fs]: Makes sense, I've changed it to trigger on multiple line rather than sequential.

I've had a go with the changelog. I didn't see an example of how duplicates are listed so added them after.

File: src/Fantomas.Core/CodePrinter.fs [hunk 3]
```diff
@@ -1401,9 +1401,12 @@ and genExpr astContext synExpr ctx =
             )
 
         | IndexWithoutDotExpr (identifierExpr, indexExpr) ->
+
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1404]: Nitpick, could you remove that first new line please?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,10 @@
 # Changelog
 
+## [Unreleased]
+
+### Fixed
+* Maintain column indentation for list of functions. [#2158](https://github.com/fsprojects/fantomas/issues/2158) duplicates: [#2195](https://github.com/fsprojects/fantomas/issues/2195), [#2201](https://github.com/fsprojects/fantomas/issues/2201), [#2242](https://github.com/fsprojects/fantomas/issues/2242), [#2392](https://github.com/fsprojects/fantomas/issues/2392)
```
Comment (nojaf) [inline CHANGELOG.md:6]: Please use the title of the issue instead of the title of the PR.
Similar to all the examples below.
And put each issue on its own line, the `duplicates:` is also somewhat strange compared to the rest.
Thanks

**Review by nojaf — CHANGES_REQUESTED** [review]
Almost there, thanks for your patience.

File: src/Fantomas.Core/CodePrinter.fs (hunk 3 above)
Comment (cgravill) [reply src/Fantomas.Core/CodePrinter.fs]: Seems like the right project for mierenneuker ;-) (my partner says that, hopefully it's light-hearted!)

Done.

File: src/Fantomas.Core/CodePrinter.fs (hunk 1 above)
Comment (cgravill) [reply src/Fantomas.Core/CodePrinter.fs]: Changed to requested style below of listing each issue

**Review by nojaf — APPROVED** [review]
Many thanks, @cgravill!
